File: delta_hedging.py
from BlackScholes import BlackScholes
from Trajectory import Trajectory

import numpy as np

########################################
############## Parameters ##############
########################################
# Market 
S0 =100
mu = 0.05
r = 0.05
sigma = 0.2

# Hhedging 
sigma_hedge = 0.2
Nhedge = 254

# Option
T = 1
K = 102

# Simulation
Nrep = 1000

St = np.array([S0]*Nrep)
dt = T/Nhedge

initialoutlay = BlackScholes(T=T,strike=K,r=r,q=0,sigma=sigma)
initialoutlay = initialoutlay.valuation(spot=S0)
Vpf = np.array([initialoutlay]* Nrep)

model = BlackScholes(T=T,strike=K,r=r,q=0,sigma=sigma_hedge,option_type='call',greek_type='delta')
a = np.array([model.valuation(s) for s in St.tolist()])
b = Vpf - a*St

from tqdm import tqdm
for i in tqdm(range(1,Nhedge)):
    St = St * np.exp((mu-0.5*sigma**2)*dt +sigma*np.sqrt(dt)*np.random.normal(size=Nrep))	
    Vpf = a * St + b * np.exp(dt*r)    
    #a <- BlackScholesFormula(St,(capT-(i-1)*dt),K, r,0,sigma_hedge,1,2)
    model = BlackScholes(T=T-(i-1)*dt,strike=K, r=r,q=0,sigma=sigma_hedge,option_type='call',greek_type='delta')
    a = np.array([model.valuation(s) for s in St.tolist()])
    b = Vpf - a * St

ST = St*np.exp((mu-0.5*sigma**2)*dt + sigma * np.sqrt(dt)*np.random.normal(size=Nrep))
Vpf = a * ST + b * np.exp(dt*r) 
hedgeerror = (Vpf-np.maximum(ST-K,0))
optionpayoff = np.maximum(ST-K,0)
  
print("Initial investment =",round(initialoutlay,4))
print("Average discounted option payoff =",round(np.exp(-r*T)*np.mean(optionpayoff),4))
print("Average discounted portfolio value =",round(np.exp(-r*T)*np.mean(Vpf),4))


import matplotlib.pyplot as plt 
plt.figure(figsize=(20,10))
plt.scatter(ST,Vpf,label="Discrete hedging of a call option",color='blue')
plt.ylabel("Value of hedge portfolio(T)")
plt.xlabel("S(T)")
x = np.sort(ST)
plt.plot(x,np.maximum(x-K,0),label='Payoff')
plt.title("Replication of vanila option")
plt.show()

File: Rscripts/Delta-Hedging.R
BlackScholesFormula  <- function (
    spot,
    timetomat,
    strike,
    r, 
    q=0, 
    sigma, 
    opttype=1, 
    greektype=1)
{ 
  
  d1<-(log(spot/strike)+ ((r-q)+0.5*sigma^2)*timetomat)/(sigma*sqrt(timetomat))
  d2<-d1-sigma*sqrt(timetomat)
  
  if (opttype==1 && greektype==1) result<-spot*exp(-q*timetomat)*pnorm(d1)-strike*exp(-r*timetomat)*pnorm(d2)
  
  if (opttype==2 && greektype==1) result<-spot*exp(-q*timetomat)*pnorm(d1)-strike*exp(-r*timetomat)*pnorm(d2)-spot*exp(-q*timetomat)+strike*exp(-r*timetomat)
  
  if (opttype==1 && greektype==2) result<-exp(-q*timetomat)*pnorm(d1)
  
  if (opttype==2 && greektype==2) result<-exp(-q*timetomat)*(pnorm(d1)-1)
  
  if (greektype==3) result<-exp(-q*timetomat)*dnorm(d1)/(spot*sigma*sqrt(timetomat))
  
  if (greektype==4) result<-exp(-q*timetomat)*spot*dnorm(d1)*sqrt(timetomat)
  
  BlackScholesFormula<-result
  
}

BlackScholesImpVol  <- function (obsprice,spot,timetomat,strike,r, q=0, opttype=1)
{ difference<- function(sigBS, obsprice,spot,timetomat,strike,r,q,opttype)
{BlackScholesFormula (spot,timetomat,strike,r,q,sigBS, opttype,1)-obsprice
}
uniroot(difference, c(10^-3,0.5),obsprice=obsprice,spot=spot,timetomat=timetomat,strike=strike,r=r,q=q,opttype=opttype,tol = .Machine$double.eps)$root
}

# INITIALIZE
# ==========

S0<-100
mu<-0.05; r<-0.05; sigma<-0.2; 

sigma_hedge<-0.2

T<-1; K<-102
capT <- T
Nhedge<-252; Nrep<-1000

opttype <- 2

# HEDGE
# =====

St<-rep(S0, length=Nrep)
dt<-T/Nhedge
#initialoutlay<-BlackScholesFormula(S0,capT,strike, r,0,sigma_hedge,1,1)
initialoutlay<-BlackScholesFormula(S0,T,K, r,0,sigma,opttype,1)

Vpf<-rep(initialoutlay,length=Nrep)

a<-BlackScholesFormula(St,T,K,r,0,sigma_hedge,opttype,2)
b<-Vpf-a*St

for(i in 2:Nhedge){
  St<-St*exp((mu-0.5*sigma^2)*dt +sigma*sqrt(dt)*rnorm(Nrep))	
  Vpf<-a*St+b*exp(dt*r)    
  a<- BlackScholesFormula(St,(capT-(i-1)*dt),K, r,0,sigma_hedge,opttype,2)
  b<-(Vpf-a*St)
}

ST<-St*exp((mu-0.5*sigma^2)*dt +sigma*sqrt(dt)*rnorm(Nrep))
Vpf<-a*ST+b*exp(dt*r) 
hedgeerror<-(Vpf-pmax(ST-K,0))
optionpayoff<-pmax(ST-K,0)

# SUMMARY STATS & GRAPHS
# ======================

print(paste("Initial investment =",round(initialoutlay,4)))
print(paste("Average discounted option payoff =",round(exp(-r*T)*mean(optionpayoff),4)))
print(paste("Average discounted portfolio value =",round(exp(-r*T)*mean(Vpf),4)))

x11()
plot(ST,Vpf,col="blue",xlab="S(T)",ylab="Value of hedge portfolio(T)",main="Discrete hedging of a call option",xlim=range(ST),ylim=range(optionpayoff))
text(min(ST),max(optionpayoff),paste("#hegde points =",Nhedge),adj=0)
points(sort(ST),pmax(K-sort(ST),0),type='l',lwd=3)